File: typescript/src/binary-search.ts
export const binarySearch = <T>(arr: T[], target: T): number => {
  let indexLow = 0;
  let indexHigh = arr.length - 1;

  while (indexLow <= indexHigh) {
    const indexMid = Math.floor((indexHigh + indexLow) / 2);
    const currentValue = arr[indexMid];

    if (currentValue === target) {
      return indexMid;
    }

    if (currentValue > target) {
      indexHigh = indexMid - 1;
    } else {
      indexLow = indexMid + 1;
    }
  }

  return -1;
};

File: go/binary_search/binary_search.go
package binary_search

import (
	"math"
)

func BinarySearch(arr []int, target int) int {
	indexHigh := len(arr) - 1
	indexLow := 0

	for indexLow < indexHigh {
		var indexMid = int(math.Floor(float64(indexHigh+indexLow) / 2))
		var currentValue = arr[indexMid]

		if currentValue == target {
			return indexMid
		}

		if currentValue > target {
			indexHigh = indexMid - 1
		} else {
			indexLow = indexMid + 1
		}
	}

	return -1
}
